File: bindings/swift/Examples/BasicUsage.swift
#!/usr/bin/env swift
/// Romcal - Basic Usage Example
///
/// This example demonstrates:
/// - Loading calendar definitions from the data folder
/// - Loading and merging resources (translations) from the data folder
/// - Creating a Romcal instance with loaded data
/// - Generating calendars with translated names
///
/// Build and run with:
///   make example
/// Or manually:
///   swift run BasicUsage

import Foundation
import Romcal

/// Get the data directory path
func getDataDir() -> URL {
    // Navigate from Examples to the data directory
    let currentDir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
    return currentDir
        .deletingLastPathComponent()  // swift
        .deletingLastPathComponent()  // bindings
        .appendingPathComponent("data")
}

/// Load all calendar definitions from the data folder
func loadCalendarDefinitions(from dataDir: URL) throws -> [[String: Any]] {
    let definitionsDir = dataDir.appendingPathComponent("definitions")
    var definitions: [[String: Any]] = []

    let fileManager = FileManager.default
    if let enumerator = fileManager.enumerator(
        at: definitionsDir,
        includingPropertiesForKeys: [.isRegularFileKey],
        options: [.skipsHiddenFiles]
    ) {
        for case let fileURL as URL in enumerator {
            if fileURL.pathExtension == "json" {
                let data = try Data(contentsOf: fileURL)
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    definitions.append(json)
                }
            }
        }
    }

    return definitions
}

/// Load all resources from the data folder
func loadResources(from dataDir: URL) throws -> [[String: Any]] {
    let resourcesDir = dataDir.appendingPathComponent("resources")
    var resources: [[String: Any]] = []

    let fileManager = FileManager.default
    var filesByLocale: [String: [URL]] = [:]

    if let enumerator = fileManager.enumerator(
        at: resourcesDir,
        includingPropertiesForKeys: [.isRegularFileKey],
        options: [.skipsHiddenFiles]
    ) {
        for case let fileURL as URL in enumerator {
            if fileURL.pathExtension == "json" {
                let locale = fileURL.deletingLastPathComponent().lastPathComponent
                if filesByLocale[locale] == nil {
                    filesByLocale[locale] = []
                }
                filesByLocale[locale]?.append(fileURL)
            }
        }
    }

    for (locale, localeFiles) in filesByLocale {
        var metadata: [String: Any]?
        var entities: [String: Any] = [:]

        for file in localeFiles {
            let data = try Data(contentsOf: file)
            if let content = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                if file.lastPathComponent == "meta.json" {
                    metadata = content["metadata"] as? [String: Any]
                } else if file.lastPathComponent.hasPrefix("entities."),
                          let fileEntities = content["entities"] as? [String: Any] {
                    entities.merge(fileEntities) { _, new in new }
                }
            }
        }

        var resource: [String: Any] = ["locale": locale]
        if let metadata = metadata {
            resource["metadata"] = metadata
        }
        if !entities.isEmpty {
            resource["entities"] = entities
        }
        resources.append(resource)
    }

    return resources
}

func main() throws {
    print("=== Romcal with Data Example ===\n")

    // Determine the data directory path
    let dataDir = getDataDir()

    guard FileManager.default.fileExists(atPath: dataDir.path) else {
        print("Error: Data directory not found at \(dataDir.path)")
        print("Please run this script from the bindings/swift directory.")
        return
    }

    // ========================================================================
    // Load Data
    // ========================================================================

    print("Loading calendar definitions...")
    let calendarDefinitions = try loadCalendarDefinitions(from: dataDir)
    print("  Loaded \(calendarDefinitions.count) calendar definitions")

    print("Loading resources...")
    let resources = try loadResources(from: dataDir)
    print("  Loaded \(resources.count) locale resources")
    print()

    // Convert to JSON strings
    let calendarDefinitionsData = try JSONSerialization.data(withJSONObject: calendarDefinitions)
    let calendarDefinitionsJson = String(data: calendarDefinitionsData, encoding: .utf8) ?? "[]"

    let resourcesData = try JSONSerialization.data(withJSONObject: resources)
    let resourcesJson = String(data: resourcesData, encoding: .utf8) ?? "[]"

    // ========================================================================
    // Create Romcal with French Calendar and Locale
    // ========================================================================

    print("Creating French calendar instance with loaded data...")
    let romcal = try RomcalCalendar(
        calendar: "france",
        locale: "fr",
        calendarDefinitionsJson: calendarDefinitionsJson,
        resourcesJson: resourcesJson
    )
    print("  Calendar: \(romcal.calendar)")
    print("  Locale: \(romcal.locale)")
    print()

    // ========================================================================
    // Generate Calendar
    // ========================================================================

    print("Generating liturgical calendar for 2026...")
    let calendar = try romcal.liturgicalCalendar(year: 2026)
    let dates = calendar.keys.sorted()
    print("  Total dates: \(dates.count)")
    print("  First date: \(dates.first ?? "N/A")")
    print("  Last date: \(dates.last ?? "N/A")")
    print()

    // ========================================================================
    // Notable Celebrations with French Names
    // ========================================================================

    print("Celebrations with French names:")

    // Easter 2026 (April 5)
    if let easter = calendar["2026-04-05"], let day = easter.first {
        print("  Easter (2026-04-05):")
        let fullname = day["fullname"] as? String ?? day["id"] as? String ?? "Unknown"
        let seasonName = day["season_name"] as? String ?? day["season"] as? String ?? "Unknown"
        let rankName = day["rank_name"] as? String ?? day["rank"] as? String ?? "Unknown"
        print("    Full name: \(fullname)")
        print("    Season: \(seasonName)")
        print("    Rank: \(rankName)")
    }

    // Assumption of Mary (August 15)
    if let assumption = calendar["2026-08-15"], let day = assumption.first {
        let fullname = day["fullname"] as? String ?? day["id"] as? String ?? "Unknown"
        print("  Assumption (2026-08-15): \(fullname)")
    }

    // All Saints (November 1)
    if let allSaints = calendar["2026-11-01"], let day = allSaints.first {
        let fullname = day["fullname"] as? String ?? day["id"] as? String ?? "Unknown"
        print("  All Saints (2026-11-01): \(fullname)")
    }

    // Christmas 2026
    if let christmas = calendar["2026-12-25"], let day = christmas.first {
        let fullname = day["fullname"] as? String ?? day["id"] as? String ?? "Unknown"
        print("  Christmas (2026-12-25): \(fullname)")
    }

    // French-specific saint: Saint Jean-Marie Vianney (August 4)
    if let vianneyDay = calendar["2026-08-04"] {
        for day in vianneyDay {
            if let id = day["id"] as? String, id.lowercased().contains("vianney") {
                let fullname = day["fullname"] as? String ?? id
                print("  St. Jean-Marie Vianney (2026-08-04): \(fullname)")
                break
            }
        }
    }

    print()

    // ========================================================================
    // Mass Calendar with French Names
    // ========================================================================

    print("Generating mass calendar for 2026...")
    let massCalendar = try romcal.massCalendar(year: 2026)
    print("  Total dates with masses: \(massCalendar.count)")
    print()

    // Christmas masses with French names
    if let christmasMasses = massCalendar["2026-12-25"] {
        print("  Christmas masses (\(christmasMasses.count) total):")
        for mass in christmasMasses {
            let massTimeName = mass["mass_time_name"] as? String ?? mass["mass_time"] as? String ?? "Unknown"
            let fullname = mass["fullname"] as? String ?? mass["id"] as? String ?? "Unknown"
            print("    - \(massTimeName): \(fullname)")
        }
    }

    print("\n=== Done ===")
}

do {
    try main()
} catch {
    print("Error: \(error)")
    exit(1)
}

File: bindings/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
Romcal - Basic Usage Example

This example demonstrates:
- Loading calendar definitions from the data folder
- Loading and merging resources (translations) from the data folder
- Creating a Romcal instance with loaded data
- Generating calendars with translated names

Run with: python examples/basic_usage.py
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from romcal import Romcal


def load_calendar_definitions(data_dir: Path) -> list[dict[str, Any]]:
    """Load all calendar definitions from the data folder."""
    definitions_dir = data_dir / "definitions"
    definitions: list[dict[str, Any]] = []

    for json_file in definitions_dir.rglob("*.json"):
        with open(json_file) as f:
            definitions.append(json.load(f))

    return definitions


def load_resources(data_dir: Path) -> list[dict[str, Any]]:
    """Load all resources from the data folder.

    Each locale has meta.json + martyrology.*.json files that need to be merged.
    """
    resources_dir = data_dir / "resources"
    resources: list[dict[str, Any]] = []

    # Group files by locale (parent directory name)
    files_by_locale: dict[str, list[Path]] = {}
    for json_file in resources_dir.rglob("*.json"):
        locale = json_file.parent.name
        if locale not in files_by_locale:
            files_by_locale[locale] = []
        files_by_locale[locale].append(json_file)

    # Merge files for each locale
    for locale, locale_files in files_by_locale.items():
        metadata: dict[str, Any] | None = None
        martyrology: dict[str, Any] = {}

        for file in locale_files:
            with open(file) as f:
                content = json.load(f)

            if file.name == "meta.json":
                metadata = content.get("metadata")
            elif file.name.startswith("martyrology.") and "martyrology" in content:
                martyrology.update(content["martyrology"])

        resources.append(
            {
                "locale": locale,
                "metadata": metadata,
                "martyrology": martyrology if martyrology else None,
            }
        )

    return resources


def main() -> None:
    print("=== Romcal with Data Example ===\n")

    # Determine the data directory path
    # This assumes running from bindings/python/
    script_dir = Path(__file__).parent
    data_dir = script_dir.parent.parent.parent / "data"

    if not data_dir.exists():
        print(f"Error: Data directory not found at {data_dir}")
        print("Please run this script from the bindings/python directory.")
        return

    # ========================================================================
    # Load Data
    # ========================================================================

    print("Loading calendar definitions...")
    calendar_definitions = load_calendar_definitions(data_dir)
    print(f"  Loaded {len(calendar_definitions)} calendar definitions")

    print("Loading resources...")
    resources = load_resources(data_dir)
    print(f"  Loaded {len(resources)} locale resources")
    print()

    # ========================================================================
    # Create Romcal with French Calendar and Locale
    # ========================================================================

    print("Creating French calendar instance with loaded data...")
    romcal = Romcal(
        calendar="france",
        locale="fr",
        calendar_definitions=calendar_definitions,
        resources=resources,
    )
    print(f"  Calendar: {romcal.calendar}")
    print(f"  Locale: {romcal.locale}")
    print()

    # ========================================================================
    # Generate Calendar
    # ========================================================================

    print("Generating liturgical calendar for 2026...")
    calendar = romcal.liturgical_calendar(2026)
    dates = sorted(calendar.keys())
    print(f"  Total dates: {len(dates)}")
    print(f"  First date: {dates[0]}")
    print(f"  Last date: {dates[-1]}")
    print()

    # ========================================================================
    # Notable Celebrations with French Names
    # ========================================================================

    print("Celebrations with French names:")

    # Easter 2026 (April 5)
    easter = calendar.get("2026-04-05")
    if easter:
        day = easter[0]
        print("  Easter (2026-04-05):")
        print(f"    Full name: {day.get('fullname', day['id'])}")
        print(f"    Season: {day.get('season_name', day['season'])}")
        print(f"    Rank: {day.get('rank_name', day['rank'])}")

    # Assumption of Mary (August 15)
    assumption = calendar.get("2026-08-15")
    if assumption:
        day = assumption[0]
        print(f"  Assumption (2026-08-15): {day.get('fullname', day['id'])}")

    # All Saints (November 1)
    all_saints = calendar.get("2026-11-01")
    if all_saints:
        day = all_saints[0]
        print(f"  All Saints (2026-11-01): {day.get('fullname', day['id'])}")

    # Christmas 2026
    christmas = calendar.get("2026-12-25")
    if christmas:
        day = christmas[0]
        print(f"  Christmas (2026-12-25): {day.get('fullname', day['id'])}")

    # French-specific saint: Saint Jean-Marie Vianney (August 4)
    vianney_day = calendar.get("2026-08-04")
    if vianney_day:
        for day in vianney_day:
            if "vianney" in day.get("id", "").lower():
                print(f"  St. Jean-Marie Vianney (2026-08-04): {day.get('fullname', day['id'])}")
                break

    print()

    # ========================================================================
    # Mass Calendar with French Names
    # ========================================================================

    print("Generating mass calendar for 2026...")
    mass_calendar = romcal.mass_calendar(2026)
    print(f"  Total dates with masses: {len(mass_calendar)}")
    print()

    # Christmas masses with French names
    christmas_masses = mass_calendar.get("2026-12-25")
    if christmas_masses:
        print(f"  Christmas masses ({len(christmas_masses)} total):")
        for mass in christmas_masses:
            mass_time_name = mass.get("mass_time_name", mass["mass_time"])
            fullname = mass.get("fullname", mass["id"])
            print(f"    - {mass_time_name}: {fullname}")

    print("\n=== Done ===")


if __name__ == "__main__":
    main()
